### PR TITLE
fix: allow transfer of `value` during contract creation

### LINKED ISSUE #682: bug: Contract with Payable Constructor doesn't deploy when value is provided
# Bug Report

**Kakarot version:**

**Current behavior:**

When a value is added to the deploy transaction on a contract with a payable constructor, the contract does not deploy and gets stuck on "pending". When the value is left as 0, the contract deploys normally.
**Expected behavior:**

The contract should deploy with a balance equal to the amount of the transaction. (specified in the "value" field in Remix).
**Steps to reproduce:**

1. Run a Katana RPC locally using Docker
2. Go to Remix and write a contract with a payable constructor:

````
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract ReceivableConstructorWithEther {
    uint public balance;

    constructor() payable {
        balance = msg.value;
    }
}
````
Compile, Deploy the contract, and run the function. The error should be:
creation of ReceivableConstructorWithEther pending...

**Related code:**

```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract ReceivableConstructorWithEther {
    uint public balance;

    constructor() payable {
        balance = msg.value;
    }
}

```

**Other information:**
For reference, I tested the functionality and got the expected behavior with the Remix Shanghai VM, and on the Goerli network.

### PR DESCRIPTION
**Pull request by bajpai244**
## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

- As of now we cannot transfer `value` when making a transaction `to`: "", i.e during an account creation.

Resolves #682 

## What is the new behavior?

- The value transfer during contract deployment is allowed.

### CODE REVIEW

File: src/kakarot/library.cairo [hunk 1]
```diff
@@ -355,23 +349,35 @@ namespace Kakarot {
...
             );
             let (return_data) = alloc();
             assert [return_data] = evm_contract_address;
             assert [return_data + 1] = starknet_contract_address;
             return (2, return_data);
         } else {
+            let (success) = transfer(origin, to, value);
+            with_attr error_message(
+                    "Kakarot: eth_call: failed to transfer {value} tokens to {to}") {
+                assert success = TRUE;
```
Comment (ftupas) [inline src/kakarot/library.cairo:379]: Not sure in Cairo0 but I think we can make `success` to be local and assert outside the if block to reduce code duplication

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py [hunk 2]
```diff
@@ -29,3 +29,12 @@ async def test_should_withdraw_call_eth(self, safe, owner, eth):
...
             ).result.balance.low == ACCOUNT_BALANCE
+
+    class TestDeploySafeWithValue:
+        async def test_deploy_safe_with_value(
+            self, safe, safe_deployer, deploy_solidity_contract
+        ):
+            safe = await deploy_solidity_contract(
+                "PlainOpcodes", "Safe", value=int("1", 16)
+            )
+            assert await safe.balance() == 1
```
Comment (ftupas) [inline tests/integration/solidity_contracts/PlainOpcodes/test_safe.py:40]:
```suggestion
            assert await safe.balance() == value
```

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py [hunk 3]
```diff
@@ -29,3 +29,12 @@ async def test_should_withdraw_call_eth(self, safe, owner, eth):
             assert (
                 await eth.balanceOf(owner.starknet_address).call()
             ).result.balance.low == ACCOUNT_BALANCE
+
+    class TestDeploySafeWithValue:
+        async def test_deploy_safe_with_value(
+            self, safe, safe_deployer, deploy_solidity_contract
+        ):
+            safe = await deploy_solidity_contract(
+                "PlainOpcodes", "Safe", value=int("1", 16)
```
Comment (ftupas) [inline tests/integration/solidity_contracts/PlainOpcodes/test_safe.py:38]:
```suggestion
            value = 1
            safe = await deploy_solidity_contract(
                "PlainOpcodes", "Safe", value=value
```

File: src/kakarot/library.cairo (hunk 1 above)
Comment (ClementWalter) [reply src/kakarot/library.cairo:379]: it may create memory holes (to be verified though)

I think it's fine for now

**Review by ClementWalter — APPROVED** [review]
I approved the suggested changes from fred to play with the buttons, let's merge

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py (hunk 3 above)
Comment (ClementWalter) [inline tests/integration/solidity_contracts/PlainOpcodes/test_safe.py:38]: please create a `value = 1`, and use it in other places in the test

Comment (bajpai244) [reply tests/integration/solidity_contracts/PlainOpcodes/test_safe.py]: done [here](https://github.com/kkrt-labs/kakarot/pull/725/commits/8bd81fe69a5f12d049c864cf98707e29c970e99b).

**Review by ClementWalter — APPROVED** [review]
don't know why fred ef test is not passing though

File: src/kakarot/library.cairo [hunk 4]
```diff
@@ -295,12 +292,7 @@ namespace Kakarot {
             bytecode_len=return_data_len,
             bytecode=return_data,
         );
-        // Increment nonce
-        IContractAccount.increment_nonce(contract_address=starknet_contract_address);
```
Comment (ftupas) [inline src/kakarot/library.cairo:299]: I think might be what's causing the ef test to fail

Comment (bajpai244) [reply src/kakarot/library.cairo:299]: Hmmm, true, this PR shouldn't touch this part of the file, might be something went wrong during rebase!

Comment (ClementWalter) [reply src/kakarot/library.cairo:299]: you should keep this indeed, good we have the test!

**Comment by ClementWalter** [comment]
So we should indeed test that
- with no constructor (or `constructor() {}` with no payable), the deployment should revert with value
- with `constructor() payable` it does gets the funds

File: src/kakarot/library.cairo [hunk 5]
```diff
@@ -252,14 +252,11 @@ namespace Kakarot {
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
-    }(origin: felt, bytecode_len: felt, bytecode: felt*) -> (
-        starknet_contract_address: felt, evm_contract_address: felt
+    }(origin: felt, evm_contract_address: felt, bytecode_len: felt, bytecode: felt*) -> (
+        starknet_contract_address: felt
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:256]: add `value` here as args to pass it to `execute`

**Review by ClementWalter — CHANGES_REQUESTED** [review]
required to have the test failing and passing as expected

File: src/kakarot/library.cairo [hunk 6]
```diff
@@ -252,14 +252,11 @@ namespace Kakarot {
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
-    }(origin: felt, bytecode_len: felt, bytecode: felt*) -> (
-        starknet_contract_address: felt, evm_contract_address: felt
-    ) {
+    }(
+        origin: felt, evm_contract_address: felt, value: felt, bytecode_len: felt, bytecode: felt*
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:256]: fix docstring plz

File: tests/integration/solidity_contracts/PlainOpcodes/Safe.sol [hunk 7]
```diff
@@ -2,6 +2,9 @@
 pragma solidity ^0.8.13;
 
 contract Safe {
+    constructor() payable{
+    }
```
Comment (ClementWalter) [inline tests/integration/solidity_contracts/PlainOpcodes/Safe.sol:6]: sol lint would make this `constructor() payable {}` you should add it to your vscode

**Review by ClementWalter — CHANGES_REQUESTED** [review]
minor fixes

File: src/kakarot/library.cairo (hunk 6 above)
Comment (bajpai244) [reply src/kakarot/library.cairo:258]: fixed

**Review by ClementWalter — APPROVED** [review]

File: src/kakarot/library.cairo (hunk 5 above)
Comment (bajpai244) [reply src/kakarot/library.cairo]: done.

**Pull request merged**